python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -744,31 +792,73 @@ def _submit_runs_and_update_backfill_in_chunks(
 
 
-def execute_asset_backfill_iteration(
-    backfill: "PartitionBackfill",
-    logger: logging.Logger,
-    workspace_process_context: IWorkspaceProcessContext,
-    instance: DagsterInstance,
-) -> Iterable[None]:
-    """Runs an iteration of the backfill, including submitting runs and updating the backfill object
-    in the DB.
+def _check_no_partitions_def_changes_to_asset(
+    asset_key: AssetKey,
+    asset_graph: AssetGraph,
+    serialized_partitions_def_id: Optional[str],
+    instance_queryer: CachingInstanceQueryer,
+) -> None:
+    if asset_key not in asset_graph.all_asset_keys:
+        raise DagsterDefinitionChangedDeserializationError(
+            f"Asset {asset_key} existed at storage-time, but no longer does"
+        )
 
-    This is a generator so that we can return control to the daemon and let it heartbeat during
-    expensive operations.
-    """
-    from dagster._core.execution.backfill import BulkActionStatus, PartitionBackfill
+    partitions_def = asset_graph.get_partitions_def(asset_key)
 
-    workspace_context = workspace_process_context.create_request_context()
-    unloadable_locations = _get_unloadable_location_names(workspace_context, logger)
-    asset_graph = ExternalAssetGraph.from_workspace(workspace_context)
+    if serialized_partitions_def_id:  # Asset was partitioned at storage time
+        if partitions_def is None:
+            raise DagsterDefinitionChangedDeserializationError(
+                f"Asset {asset_key} had a PartitionsDefinition at storage-time, but no longer"
+                " does"
+            )
 
-    if backfill.serialized_asset_backfill_data is None:
-        check.failed("Asset backfill missing serialized_asset_backfill_data")
+        if (
+            partitions_def.get_serializable_unique_identifier(instance_queryer)
+            != serialized_partitions_def_id
+        ):
+            raise DagsterDefinitionChangedDeserializationError(
+                f"This partitions definition for asset {asset_key} has changed since this backfill"
+                " was stored"
+            )
+
+    else:  # Asset unpartitioned at storage time
+        if partitions_def is not None:
+            raise DagsterDefinitionChangedDeserializationError(
+                f"Asset {asset_key} was not partitioned at storage-time, but is now"
+            )
+
+
+def _check_and_deserialize_asset_backfill_data(
+    workspace_context: BaseWorkspaceRequestContext,
+    backfill: "PartitionBackfill",
+    asset_graph: AssetGraph,
+    instance_queryer: CachingInstanceQueryer,
+    logger: logging.Logger,
+) -> Optional[AssetBackfillData]:
+    unloadable_locations = _get_unloadable_location_names(workspace_context, logger)
 
     try:
-        previous_asset_backfill_data = AssetBackfillData.from_serialized(
-            backfill.serialized_asset_backfill_data, asset_graph, backfill.backfill_timestamp
-        )
+        if backfill.serialized_asset_backfill_data:
```

Comment:
Could it make sense to consolidate parts of this implementation with `get_asset_backfill_data`?